File: Nickvision.Application.Shared/Controllers/MainWindowController.cs
using Microsoft.Extensions.Logging;
using Nickvision.Application.Shared.Helpers;
using Nickvision.Application.Shared.Models;
using Nickvision.Application.Shared.Services;
using Nickvision.Desktop.Application;
using Nickvision.Desktop.Globalization;
using Nickvision.Desktop.Network;
using Nickvision.Desktop.Notifications;
using System;
using System.Threading.Tasks;

namespace Nickvision.Application.Shared.Controllers;

public class MainWindowController
{
    ILogger<MainWindowController> _logger;
    private readonly AppInfo _appInfo;
    private readonly IConfigurationService _configurationService;
    private readonly IFolderService _folderService;
    private readonly INotificationService _notificationService;
    private readonly ITranslationService _translationService;
    private readonly IUpdaterService _updaterService;
    private AppVersion _latestVersion;

    public MainWindowController(ILogger<MainWindowController> logger, AppInfo appInfo, IArgumentsService argumentsService, IConfigurationService configurationService, IFolderService folderService, INotificationService notificationService, ITranslationService translationService, IUpdaterService updaterService)
    {
        _logger = logger;
        _appInfo = appInfo;
        _configurationService = configurationService;
        _folderService = folderService;
        _notificationService = notificationService;
        _translationService = translationService;
        _updaterService = updaterService;
        _latestVersion = appInfo.Version!;
        _translationService.Language = _configurationService.TranslationLanguage;
        _logger.LogInformation($"Received command-line arguments: [{string.Join(", ", argumentsService.Data)}]");
        // Translate strings
        _appInfo.ShortName = _translationService._("Application");
        _appInfo.Description = _translationService._("Create new Nickvision applications.");
        _appInfo.ExtraLinks.Add(_translationService._("Matrix Chat"), new Uri("https://matrix.to/#/#nickvision:matrix.org"));
        _appInfo.Developers.Add(_translationService._("Nicholas Logozzo"), "https://github.com/nlogozzo");
        _appInfo.Developers.Add(_translationService._("Contributors on GitHub ❤️"), "https://github.com/NickvisionApps/Application/graphs/contributors");
        _appInfo.Designers.Add(_translationService._("Nicholas Logozzo"), "https://github.com/nlogozzo");
        _appInfo.Designers.Add(_translationService._("Fyodor Sobolev"), "https://github.com/fsobolev");
        _appInfo.Designers.Add(_translationService._("DaPigGuy"), "https://github.com/DaPigGuy");
        _appInfo.Artists.Add(_translationService._("David Lapshin"), "https://github.com/daudix");
        _appInfo.TranslationCredits = _translationService._("translation-credits");
    }

    public bool CanShutdown => true;

    public string? CurrentFolder => _folderService.Path;

    public string Greeting => DateTime.Now.Hour switch
    {
        >= 0 and < 6 => _translationService._p("Night", "Good Morning!"),
        < 12 => _translationService._p("Morning", "Good Morning!"),
        < 18 => _translationService._("Good Afternoon!"),
        < 24 => _translationService._("Good Evening!"),
        var _ => _translationService._("Good Day!")
    };

    public Theme Theme => _configurationService.Theme;

    public WindowGeometry WindowGeometry
    {
        get => _configurationService.WindowGeometry;

        set
        {
            _configurationService.WindowGeometry = value;
            _configurationService.Save();
        }
    }

    public async Task CheckForUpdatesAsync(bool showNotificationForNoUpdates)
    {
        _logger.LogInformation("Checking for updates...");
        var stableVersion = await _updaterService.GetLatestStableVersionAsync();
        if (stableVersion is not null)
        {
            _latestVersion = stableVersion;
        }
        if (_configurationService.AllowPreviewUpdates)
        {
            var previewVersion = await _updaterService.GetLatestPreviewVersionAsync();
            if (previewVersion is not null && previewVersion > stableVersion)
            {
                _latestVersion = previewVersion;
            }
        }
        if (_latestVersion > _appInfo.Version!)
        {
            if (!OperatingSystem.IsLinux())
            {
                _logger.LogInformation($"New application update available: {_latestVersion}");
                _notificationService.Send(new AppNotification(_translationService._("New {0} update available: {1}", _appInfo.ShortName!, _latestVersion.ToString()), NotificationSeverity.Success)
                {
                    Action = "update"
                });
            }
        }
        else
        {
            _logger.LogInformation("No application update available.");
            if (showNotificationForNoUpdates)
            {
                _notificationService.Send(new AppNotification(_translationService._("No update available"), NotificationSeverity.Warning));
            }
        }
    }

    public string GetDebugInformation(string extraInformation = "") => Desktop.System.Environment.GetDebugInformation(_appInfo, extraInformation);

    public void OpenFolder(string path) => _folderService.Open(path);

    public void CloseFolder() => _folderService.Close();

    public async Task WindowsUpdateAsync(IProgress<DownloadProgress> progress)
    {
        var res = await _updaterService.WindowsApplicationUpdateAsync(_latestVersion, progress);
        if (res)
        {
            _notificationService.Send(new AppNotification(_translationService._("Starting {0} installer...", _appInfo.ShortName!), NotificationSeverity.Success));
        }
        else
        {
            _notificationService.Send(new AppNotification(_translationService._("Unable to download and install the update"), NotificationSeverity.Error));
        }
    }
}

File: libapplication/src/controllers/main_window_controller.cpp
#include "controllers/main_window_controller.h"
#include <chrono>
#include <format>
#include <optional>

using namespace application::services;
using namespace application::models;
using namespace desktop::app;
using namespace desktop::network;
using namespace desktop::notifications;
using namespace desktop::system;
using namespace desktop::updates;

namespace application::controllers
{
	main_window_controller::main_window_controller(std::shared_ptr<app_info> app_info, std::shared_ptr<app_config_service> app_config_service,
	                                               std::shared_ptr<folder_service> folder_service, std::shared_ptr<notification_service> notification_service,
	                                               std::shared_ptr<translation_service> translation_service, std::shared_ptr<update_service> update_service)
	    : m_app_info{ std::move(app_info) },
	      m_app_config_service{ std::move(app_config_service) },
	      m_folder_service{ std::move(folder_service) },
	      m_notification_service{ std::move(notification_service) },
	      m_translation_service{ std::move(translation_service) },
	      m_update_service{ std::move(update_service) }
	{
		m_app_info->set_short_name(m_translation_service->_("Application"));
		m_app_info->set_description(m_translation_service->_("Create new Nickvision application"));
		m_app_info->add_extra_link(m_translation_service->_("Matrix Chat"), "https://matrix.to/#/#nickvision:matrix.org");
		m_app_info->add_developer(m_translation_service->_("Nicholas Logozzo"), "https://githjub.com/nlogozzo");
		m_app_info->add_developer(m_translation_service->_("Contributors on GitHub ❤️"), "https://github.com/NickvisionApps/Application/graphs/contributors");
		m_app_info->add_designer(m_translation_service->_("Nicholas Logozzo"), "https://github.com/nlogozzo");
		m_app_info->add_designer(m_translation_service->_("Fyodor Sobolev"), "https://github.com/fsobolev");
		m_app_info->add_designer(m_translation_service->_("DaPigGuy"), "https://github.com/DaPigGuy");
		m_app_info->add_artist(m_translation_service->_("David Lapshin"), "https://github.com/daudix");
		m_app_info->set_translation_credits(m_translation_service->_("translation-credits"));
	}

	bool main_window_controller::can_shutdown() const
	{
		return true;
	}

	std::shared_ptr<app_info> main_window_controller::get_app_info() const
	{
		return m_app_info;
	}

	const std::filesystem::path& main_window_controller::get_current_folder() const
	{
		return m_folder_service->get_path();
	}

	std::string main_window_controller::get_greeting() const
	{
		long hours{ std::chrono::duration_cast<std::chrono::hours>(std::chrono::system_clock::now().time_since_epoch()).count() % 24 };
		if (hours >= 0 && hours < 6)
		{
			return m_translation_service->_p("Night", "Good Morning!");
		}
		if (hours < 12)
		{
			return m_translation_service->_p("Morning", "Good Morning!");
		}
		if (hours < 18)
		{
			return m_translation_service->_("Good Afternoon!");
		}
		if (hours < 24)
		{
			return m_translation_service->_("Good Evening!");
		}
		return m_translation_service->_("Good Day!");
	}

	theme main_window_controller::get_theme() const
	{
		return m_app_config_service->get_theme();
	}

	desktop::app::window_geometry main_window_controller::get_window_geometry() const
	{
		return m_app_config_service->get_window_geometry();
	}

	void main_window_controller::set_window_geometry(const desktop::app::window_geometry& geometry)
	{
		m_app_config_service->set_window_geometry(geometry);
	}

	std::string main_window_controller::get_debugging_information() const
	{
		std::string info{ std::format("{} {}\n\n", m_app_info->get_short_name(), m_app_info->get_version().str()) };
		info += environment::get_debugging_information();
		return info;
	}

	void main_window_controller::queue_check_for_updates(bool show_no_update_notification)
	{
		std::optional<version> stable{ m_update_service->get_latest_version(false) };
		if (stable.has_value())
		{
			m_latest_version = *stable;
		}
		if (m_app_config_service->allow_preview_updates())
		{
			std::optional<version> preview{ m_update_service->get_latest_version(true) };
			if (preview.has_value() && stable.has_value() && *preview > *stable)
			{
				m_latest_version = *preview;
			}
		}
		if (m_latest_version > m_app_info->get_version())
		{
#ifdef __linux__
			return;
#else
			m_notification_service->send(
			    app_notification{ m_translation_service->_("New {} update available: {}", m_app_info->get_short_name(), m_latest_version.str()),
			                      notification_severity::information, "update" });
#endif
		}
		else if (show_no_update_notification)
		{
			m_notification_service->send(app_notification{ m_translation_service->_("No Update Available"), notification_severity::warning });
		}
	}

#ifdef _WIN32
	void main_window_controller::install_update_for_windows(const std::function<void(const desktop::network::download_progress&)>& progress)
	{
		bool result{ m_update_service->install_update_for_windows(m_latest_version, progress) };
		if (result)
		{
			m_notification_service->send(
			    app_notification{ m_translation_service->_("Starting {} installer...", m_app_info->get_short_name()), notification_severity::success });
		}
		else
		{
			m_notification_service->send(
			    app_notification{ m_translation_service->_("Unable to download and install the update"), notification_severity::error });
		}
	}
#endif

	void main_window_controller::open_folder(const std::filesystem::path& path)
	{
		m_folder_service->open(path);
	}

	void main_window_controller::close_folder()
	{
		m_folder_service->close();
	}
}